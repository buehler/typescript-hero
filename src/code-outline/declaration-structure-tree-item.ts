import {
  ClassDeclaration,
  Declaration,
  FunctionDeclaration,
  GetterDeclaration,
  InterfaceDeclaration,
  MethodDeclaration,
  PropertyDeclaration,
  SetterDeclaration,
  VariableDeclaration,
} from 'typescript-parser';
import { stringTemplate } from 'typescript-parser/utilities/StringTemplate';
import { CompletionItemKind, ExtensionContext, TreeItemCollapsibleState } from 'vscode';

import { getItemKind } from '../utilities/utility-functions';
import BaseStructureTreeItem from './base-structure-tree-item';

const fileTemplate = stringTemplate`./src/assets/icons/declarations/${0}.svg`;

/**
 * Function to calculate the displayed name of the declaration structure item.
 *
 * @param {Declaration} declaration
 * @returns {string}
 */
function getDeclarationLabel(declaration: Declaration): string {
  if (
    declaration instanceof FunctionDeclaration ||
    declaration instanceof MethodDeclaration
  ) {
    const params = declaration.parameters.map(p => p.name + (p.type ? `: ${p.type}` : '')).join(', ');
    return `${declaration.name}(${params})${declaration.type ? `: ${declaration.type}` : ''}`;
  }

  if (declaration instanceof PropertyDeclaration) {
    return declaration.name + (declaration.type ? `: ${declaration.type}` : '');
  }

  if (
    declaration instanceof ClassDeclaration ||
    declaration instanceof InterfaceDeclaration
  ) {
    return declaration.name + (declaration.typeParameters ? `<${declaration.typeParameters.join(', ')}>` : '');
  }

  if (
    declaration instanceof GetterDeclaration ||
    declaration instanceof SetterDeclaration
  ) {
    return `${declaration instanceof GetterDeclaration ? 'get' : 'set'}() ${declaration.name}` +
      `${declaration.type ? `: ${declaration.type}` : ''}`;
  }

  return declaration.name;
}

/**
 * Structure item that represents a typescript declaration of any way.
 *
 * @export
 * @class DeclarationStructureTreeItem
 * @extends {BaseStructureTreeItem}
 */
export default class DeclarationStructureTreeItem extends BaseStructureTreeItem {
  public get iconPath(): string | undefined {
    switch (getItemKind(this.declaration)) {
      case CompletionItemKind.Class:
      case CompletionItemKind.Keyword:
        return this.context.asAbsolutePath(fileTemplate('class'));
      case CompletionItemKind.Interface:
        return this.context.asAbsolutePath(fileTemplate('interface'));
      case CompletionItemKind.Enum:
        return this.context.asAbsolutePath(fileTemplate('enum'));
      case CompletionItemKind.Function:
      case CompletionItemKind.Method:
        return this.context.asAbsolutePath(fileTemplate('callable'));
      case CompletionItemKind.Module:
        return this.context.asAbsolutePath(fileTemplate('module'));
      case CompletionItemKind.Property:
        return this.context.asAbsolutePath(fileTemplate('property'));
      default:
        break;
    }

    if (getItemKind(this.declaration) === CompletionItemKind.Variable) {
      return (this.declaration as VariableDeclaration).isConst ?
        this.context.asAbsolutePath(fileTemplate('const')) :
        this.context.asAbsolutePath(fileTemplate('variable'));
    }

    return this.context.asAbsolutePath(fileTemplate('default'));
  }

  constructor(public declaration: Declaration, private context: ExtensionContext) {
    super(getDeclarationLabel(declaration));

    if (
      declaration instanceof ClassDeclaration ||
      declaration instanceof InterfaceDeclaration
    ) {
      this.collapsibleState = TreeItemCollapsibleState.Collapsed;
    } else {
      this.command = this.command = this.createJumpToCommand([declaration]);
    }
  }

  public getChildren(): BaseStructureTreeItem[] {
    console.log('DIIINI muETER');
    if (
      this.declaration instanceof ClassDeclaration ||
      this.declaration instanceof InterfaceDeclaration
    ) {
      return [
        ...this.declaration.accessors.map(p => new DeclarationStructureTreeItem(p, this.context)),
        ...this.declaration.properties.map(p => new DeclarationStructureTreeItem(p, this.context)),
        ...this.declaration.methods.map(m => new DeclarationStructureTreeItem(m, this.context)),
      ];
    }
    return [];
  }
}

