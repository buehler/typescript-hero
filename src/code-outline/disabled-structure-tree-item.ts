import BaseStructureTreeItem from './base-structure-tree-item';

/**
 * Structure tree item that is used when the feature is disabled via config.
 *
 * @export
 * @class DisabledStructureTreeItem
 * @extends {BaseStructureTreeItem}
 */
export default class DisabledStructureTreeItem extends BaseStructureTreeItem {
  constructor() {
    super('Feature is disabled.');
  }
}
