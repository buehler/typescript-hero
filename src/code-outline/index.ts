import { inject, injectable } from 'inversify';
import { Subscription } from 'rxjs';
import { File, TypescriptParser } from 'typescript-parser';
import {
  Disposable,
  Event,
  EventEmitter,
  ExtensionContext,
  ProviderResult,
  TreeDataProvider,
  window,
  workspace,
} from 'vscode';

import Activatable from '../activatable';
import Configuration from '../configuration';
import iocSymbols from '../ioc-symbols';
import { Logger } from '../utilities/Logger';
import { getScriptKind } from '../utilities/utilityFunctions';
import BaseStructureTreeItem from './base-structure-tree-item';
import DeclarationStructureTreeItem from './declaration-structure-tree-item';
import DisabledStructureTreeItem from './disabled-structure-tree-item';
import { ImportsStructureTreeItem } from './imports-structure-tree-item';
import NotParseableStructureTreeItem from './not-parseable-structure-tree-item';
import ResourceStructureTreeItem from './resource-structure-tree-item';

@injectable()
export default class CodeOutline implements Activatable, TreeDataProvider<BaseStructureTreeItem> {
  private _onDidChangeTreeData: EventEmitter<BaseStructureTreeItem | undefined>;

  private subscription: Subscription;
  private disposables: Disposable[] = [];
  private documentCache?: File;

  public get onDidChangeTreeData(): Event<BaseStructureTreeItem | undefined> {
    return this._onDidChangeTreeData.event;
  }

  constructor(
    @inject(iocSymbols.extensionContext) private context: ExtensionContext,
    @inject(iocSymbols.logger) private logger: Logger,
    @inject(iocSymbols.configuration) private config: Configuration,
    @inject(iocSymbols.parser) private parser: TypescriptParser,
  ) { }

  public setup(): void {
    this.logger.debug('Setting up DocumentOutline.');
    this.subscription = this.config.configurationChanged.subscribe(() => {
      if (this.config.codeOutline.isEnabled() && !this.disposables) {
        this.start();
      } else if (!this.config.codeOutline.isEnabled() && this.disposables) {
        this.stop();
      }
    });
  }

  public start(): void {
    if (!this.config.codeOutline.isEnabled()) {
      this.logger.debug(`Not starting DocumentOutline. It's disabled by config.`);
      return;
    }
    this.logger.debug('Starting up DocumentOutline.');
    this._onDidChangeTreeData = new EventEmitter<BaseStructureTreeItem | undefined>();
    this.disposables.push(window.registerTreeDataProvider('codeOutline', this));
    this.context.subscriptions.push(this._onDidChangeTreeData);
    this.context.subscriptions.push(window.onDidChangeActiveTextEditor(() => this.activeWindowChanged()));
    this.context.subscriptions.push(workspace.onDidSaveTextDocument(() => this.activeWindowChanged()));
  }

  public stop(): void {
    if (this.config.codeOutline.isEnabled()) {
      this.logger.debug(`Not stopping DocumentOutline. It's enabled by config.`);
      return;
    }
    this.logger.debug('Stopping DocumentOutline.');
    for (const disposable of this.disposables) {
      disposable.dispose();
    }
    this.disposables = [];
  }

  public dispose(): void {
    this.logger.debug('Disposing DocumentOutline.');
    if (this.subscription) {
      this.subscription.unsubscribe();
      delete this.subscription;
    }
    for (const disposable of this.disposables) {
      disposable.dispose();
    }
    this.disposables = [];
  }

  public getTreeItem(element: BaseStructureTreeItem): BaseStructureTreeItem {
    return element;
  }

  public async getChildren(element?: BaseStructureTreeItem): Promise<ProviderResult<BaseStructureTreeItem[]>> {
    if (!window.activeTextEditor) {
      return [];
    }

    if (!this.config.codeOutline.isEnabled()) {
      return [new DisabledStructureTreeItem()];
    }

    if (!this.config.parseableLanguages().some(
      lang => lang === window.activeTextEditor!.document.languageId,
    )) {
      return [new NotParseableStructureTreeItem()];
    }

    if (!this.documentCache) {
      try {
        this.documentCache = await this.parser.parseSource(
          window.activeTextEditor.document.getText(),
          getScriptKind(window.activeTextEditor.document.fileName),
        );
      } catch (e) {
        this.logger.error(
          `[DocumentOutline] document could not be parsed, error: ${e}`,
        );
        return [];
      }
    }

    if (!element) {
      const items: BaseStructureTreeItem[] = [];
      if (this.documentCache.imports && this.documentCache.imports.length) {
        items.push(new ImportsStructureTreeItem(this.documentCache, this.context));
      }
      items.push(...this.documentCache.resources.map(r => new ResourceStructureTreeItem(r, this.context)));
      items.push(...this.documentCache.declarations.map(d => new DeclarationStructureTreeItem(d, this.context)));
      return items;
    }
    return element.getChildren();
  }

  /**
     * Method that recalculates the current document when the active window changed.
     *
     * @private
     *
     * @memberof DocumentSymbolStructureExtension
     */
  private activeWindowChanged(): void {
    this.logger.debug('[DocumentOutline] activeWindowChanged, reparsing');
    this.documentCache = undefined;
    this._onDidChangeTreeData.fire();
  }
}
