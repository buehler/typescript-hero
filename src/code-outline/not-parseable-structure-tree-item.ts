import BaseStructureTreeItem from './base-structure-tree-item';

/**
 * Structure tree item that is used when a document is not parseable (i.e. not typescript)
 *
 * @export
 * @class NotParseableStructureTreeItem
 * @extends {BaseStructureTreeItem}
 */
export default class NotParseableStructureTreeItem extends BaseStructureTreeItem {
  constructor() {
    super('File not parseable.');
  }
}
