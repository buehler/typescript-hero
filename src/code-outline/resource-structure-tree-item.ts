import { Resource } from 'typescript-parser';
import { ExtensionContext, TreeItemCollapsibleState } from 'vscode';

import BaseStructureTreeItem from './base-structure-tree-item';
import DeclarationStructureTreeItem from './declaration-structure-tree-item';
import { ImportsStructureTreeItem } from './imports-structure-tree-item';

/**
 * Structure item that represents an additional resource in a file (namespace etc).
 *
 * @export
 * @class ResourceStructureTreeItem
 * @extends {BaseStructureTreeItem}
 */
export default class ResourceStructureTreeItem extends BaseStructureTreeItem {
  constructor(public resource: Resource, private context: ExtensionContext) {
    super(resource.identifier);
    this.collapsibleState = TreeItemCollapsibleState.Collapsed;
    this.iconPath = this.context.asAbsolutePath('./src/extension/assets/icons/declarations/module.svg');
  }

  public getChildren(): BaseStructureTreeItem[] {
    const items: BaseStructureTreeItem[] = [];
    if (this.resource.imports && this.resource.imports.length) {
      items.push(new ImportsStructureTreeItem(this.resource, this.context));
    }
    items.push(...this.resource.resources.map(r => new ResourceStructureTreeItem(r, this.context)));
    items.push(...this.resource.declarations.map(d => new DeclarationStructureTreeItem(d, this.context)));
    return items;
  }
}
