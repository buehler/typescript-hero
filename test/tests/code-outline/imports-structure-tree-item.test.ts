import { expect } from 'chai';
import { ExternalModuleImport, File, NamedImport, NamespaceImport, StringImport, SymbolSpecifier } from 'typescript-parser';
import { ExtensionContext } from 'vscode';

import { ImportsStructureTreeItem, ImportStructureTreeItem } from '../../../src/code-outline/imports-structure-tree-item';
import ioc from '../../../src/ioc';
import iocSymbols from '../../../src/ioc-symbols';

describe('ImportsStructureTreeItem', () => {

  let context: ExtensionContext;

  before(() => {
    context = ioc.get<ExtensionContext>(iocSymbols.extensionContext);
  });

  it('should create a tree item', () => {
    const resource = new File('./path', '/root', 0, 100);
    resource.imports.push(new NamedImport('lib', 0, 1));
    const item = new ImportsStructureTreeItem(resource, context);

    expect(item).to.exist;
  });

  it('should return the children for the imports', () => {
    const resource = new File('./path', '/root', 0, 100);
    resource.imports.push(new NamedImport('lib', 0, 1));
    const item = new ImportsStructureTreeItem(resource, context);

    // snapshotitem.getChildren());
  });

});

describe('ImportStructureTreeItem', () => {

  let context: ExtensionContext;

  before(() => {
    context = ioc.get<ExtensionContext>(iocSymbols.extensionContext);
  });

  it('should create a tree item', () => {
    const item = new ImportStructureTreeItem(new NamedImport('lib', 0, 1), context);

    expect(item).to.exist;
  });

  it('should return the correct children for the imports', () => {
    const testFn = (imp) => {
      const item = new ImportStructureTreeItem(imp, context);
      return item.getChildren();
    };

    const namedImp = new NamedImport('named-imp', 0, 1);
    const stringImp = new StringImport('str-imp');
    const extImp = new ExternalModuleImport('ext-imp', 'extImp', 0, 1);
    const namespaceImp = new NamespaceImport('namespace-imp', 'namespace', 0, 1);
    const specImp = new NamedImport('named-spec-imp', 0, 1);
    specImp.defaultAlias = 'default';
    specImp.specifiers.push(new SymbolSpecifier('spec'));

    // snapshottestFn, namedImp, stringImp, extImp, namespaceImp, specImp);
  });

});
