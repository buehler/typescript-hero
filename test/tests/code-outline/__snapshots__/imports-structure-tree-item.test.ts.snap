// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportsStructureTreeItem should return the children for the imports 1`] = `
Array [
  ImportStructureTreeItem {
    "collapsibleState": 1,
    "command": Object {
      "arguments": Array [
        NamedImport {
          "end": 1,
          "libraryName": "lib",
          "specifiers": Array [],
          "start": 0,
        },
      ],
      "command": "typescriptHero.codeOutline.gotoNode",
      "title": "Jump to node",
    },
    "context": ContextMock {
      "extensionPath": "",
      "storagePath": "",
      "subscriptions": Array [],
    },
    "iconPath": "./src/assets/icons/declarations/import.svg",
    "label": "lib",
    "tsImport": NamedImport {
      "end": 1,
      "libraryName": "lib",
      "specifiers": Array [],
      "start": 0,
    },
  },
]
`;
